docs/v3.x/guides/unit-testing.md

```diff
@@ -0,0 +1,340 @@
+
+```json
+  "scripts": {
+    "develop": "strapi develop",
+    "start": "strapi start",
+    "build": "strapi build",
+    "strapi": "strapi",
+    "test": "jest --forceExit --detectOpenHandles"
+  },
+```
+
+and add those line at the bottom of file
+
+```json
+  "jest": {
+    "testPathIgnorePatterns": [
+      "/node_modules/",
+      ".tmp",
+      ".cache"
+    ],
+    "testEnvironment": "node"
+  }
+```
+
+Those will inform `Jest` not to look for test inside the folder where it shouldn't.
+::::
+
+## Introduction
+
+### Testing environment
+
+Test framework must have a clean empty environment to perform valid test and also not to interfere with current database.
+
+Once `jest` is running it uses the `test` [enviroment](../concepts/configurations.md#environments) (switching `NODE_ENV` to `test`)
+so we need to create a special environment setting for this purpose.
+Create a new config for test env `./config/env/test/database.json` and add the following value `"filename": ".tmp/test.db"` - the reason of that is that we want to have a separate sqlite database for tests, so our test will not touch real data.
+This file will be temporary, each time test is finished, we will remove that file that every time tests are run on the clean database.
+The whole file will look like this:
+
+**Path —** `./config/env/test/database.json`
+
+```json
+{
+  "defaultConnection": "default",
+  "connections": {
+    "default": {
+      "connector": "bookshelf",
+      "settings": {
+        "client": "sqlite",
+        "filename": ".tmp/test.db"
+      },
+      "options": {
+        "useNullAsDefault": true,
+        "pool": {
+          "min": 0,
+          "max": 15
+        }
+      }
+    }
+  }
+}
+```
+
+### Strapi instance
+
+In order to test anything we need to have a strapi instance that runs in the testing eviroment,
+bascially we want to get instance of strapi app as object, similar like creating an instance for [process manager](process-manager).
+
+These tasks require adding some files - let's create a folder `tests` where all the tests will be put and inside it, next to folder `helpers` where main Strapi helper will be in file strapi.js.
+
+**Path —** `./tests/helpers/strapi.js`
+
+```js
+const Strapi = require('strapi');
+const http = require('http');
+
+let instance;
+
+async function setupStrapi() {
+  if (!instance) {
+    /** the following code in copied from `./node_modules/strapi/lib/Strapi.js` */
+    await Strapi().load();
+    instance = strapi; // strapi is global now
+    await instance.app
+      .use(instance.router.routes()) // populate KOA routes
+      .use(instance.router.allowedMethods()); // populate KOA methods
+
+    instance.server = http.createServer(instance.app.callback());
+  }
+  return instance;
+}
+module.exports = { setupStrapi };
+```
+
+### Test strapi instance
+
+We need a main entry file for our tests, one that will also test our helper file.
+
+**Path —** `./tests/helpers/strapi.js`
```

Comment:
I think this needs to be changed to `./tests/app.test.js`

Reply:
```suggestion
**Path —** `./tests/app.test.js`
```